geomesa-kafka-bundle/geomesa-kafka-processors/src/main/scala/org/geomesa/nifi/processors/kafka/GetGeoMesaKafkaRecord.scala

```diff
@@ -139,95 +134,111 @@ class GetGeoMesaKafkaRecord extends AbstractProcessor {
-      val record = readNextRecord()
-      if (record != null) {
-        val count = writer.write(record).getRecordCount
-        if (count < maxBatchSize) {
-          writeRemaining(writer, timeout, count)
+      val errors = results.collect { case Failure(e) => e }
+      if (errors.nonEmpty) {
+        try {
+          val e = errors.head
+          errors.tail.foreach(e.addSuppressed)
+          logger.error("Error processing message batch:", e)
+          results.collect { case Success(f) => Try(session.remove(f)) }
+        } finally {
+          processor.done.put(false)
         }
-      } else if (count < minBatchSize) {
-        writeRemaining(writer, timeout, count)
+      } else {
+        var ok = true
+        results.foreach { case Success(f) =>
+          try {
+            if (f.getAttribute("record.count").toInt <= 0) {
+              logger.debug("Removing flow file, no records were written")
+              session.remove(f)
+            } else {
+              session.transfer(f, SuccessRelationship)
+            }
+          } catch {
+            case NonFatal(e) =>
+              ok = false
+              logger.error(s"Error transferring flow file $f:", e)
+              Try(session.remove(f))
+          }
+        }
+        processor.done.put(ok)
       }
     }
   }
 
-  private def readNextRecord(): Record = {
-    val feature = queue.poll(pollTimeout, TimeUnit.MILLISECONDS)
-    if (feature == null) { null } else {
-      converter.convert(feature)
-    }
-  }
-
   @OnRemoved
   @OnStopped
   @OnShutdown
   def cleanup(): Unit = {
     logger.info("Processor shutting down")
     val start = System.currentTimeMillis()
-    if (ds != null) {
-      fs.removeFeatureListener(listener)
-      CloseWithLogging(ds)
-      fs = null
-      ds = null
+    if (consumer != null) {
+      CloseWithLogging(consumer)
+      consumer = null
     }
     logger.info(s"Shut down in ${System.currentTimeMillis() - start}ms")
   }
 
-  class RecordFeatureListener extends FeatureListener {
-    override def changed(event: FeatureEvent): Unit = {
-      event match {
-        case e: KafkaFeatureChanged => queue.put(e.feature)
-        case _: KafkaFeatureRemoved => // no-op
-        case _: KafkaFeatureCleared => // no-op
-        case e                      => logger.error(s"Unknown event $e")
+  class RecordProcessor extends GeoMessageProcessor {
+
+    val ready = new SynchronousQueue[Seq[SimpleFeature]]()
+    val done = new SynchronousQueue[Boolean]()
+
+    private var lastSuccess = System.currentTimeMillis()
+
+    override def consume(records: Seq[GeoMessage]): Boolean = {
+      val features = records.collect { case GeoMessage.Change(f) => f }
+      if (features.size < minBatchSize && maxLatencyMillis.forall(_ > System.currentTimeMillis() - lastSuccess)) {
+        logger.debug(s"Received ${features.size} records but waiting for larger batch")
+        false
+      } else if (!ready.offer(features, 10, TimeUnit.SECONDS)) {
```

Comment:
What's this condition?  

I hate to ask, but should the 10 seconds be configurable?  (E.g., would this affect data latency for a low volume and/or bursty feed?

Reply:
so this code is slightly complicated because we have 2 asynchronous callbacks that we have to coordinate - the onTrigger method and the kafka consume method. this is the amount of time that the consume method will wait for onTrigger to be invoked before bailing and giving a chance to read more messages off the kafka queue. there are 5 nifi configurations that you could tweak here - minBatchSize and maxLatency to control to the number of messages to wait for, the processor poll duration, which is how long onTrigger will wait for the consumer and defaults to 1 second, and then the nifi yield duration and poll duration, which are by default 30 seconds and 1 second, respectively. additionally, you can configure the kafka batch size and poll duration, which will control how often the consumer gets invoked. so, I could see some configuration combinations along with some kafka message time patterns that result in messages being delayed for a bit, but there are already a lot of levers to pull, so i'm not sure if making this configurable would be helpful or not.

Reply:
the default yield is only 1 second, not 30, sorry. 30 is if the processor errors out with an uncaught exception.